crates/typst-utils/src/version.rs

```diff
@@ -0,0 +1,90 @@
+//! Typst version information.
+
+/// Returns the version of Typst.
+///
+/// The information is read from the following sources in order of precedence:
+///
+/// - For the version number:
+///  - The `TYPST_VERSION` environment variable
+///  - The `CARGO_PKG_VERSION` environment variable
+///
+/// - For the commit sha:
+///   - The `TYPST_COMMIT_SHA` environment variable
+///   - The output of `git rev-parse HEAD`
+///
+/// # Panics
+/// If the `TYPST_VERSION` environment variable holds a version string that
+/// doesn't conform to SemVer.
+pub fn version() -> TypstVersion {
+    *crate::singleton!(TypstVersion, {
+        let raw = env!("TYPST_VERSION");
+        let commit = option_env!("TYPST_COMMIT_SHA");
+        match semver::Version::parse(raw) {
+            Ok(version) => {
+                return TypstVersion {
+                    major: version.major.try_into().unwrap(),
+                    minor: version.minor.try_into().unwrap(),
+                    patch: version.patch.try_into().unwrap(),
+                    raw,
+                    commit,
+                };
+            }
+            Err(err) => {
+                panic!("failed to parse {raw:?} as semantic version number: {err:?}")
+            }
+        }
+    })
+}
+
+/// Typst version definition.
+///
+/// This structure contains the current Typst version. To query the precise
+/// version number, refer to the [`TypstVersion::major()`],
+/// [`TypstVersion::minor()`] and [`TypstVersion::patch()`] functions. You can
+/// read the underlying, raw version string (e.g. for CLI output) with
+/// [`TypstVersion::raw`].
+///
+/// Optionally, this may also contain the Git commit from which Typst was built.
+/// However, this field may be unpopulated.
```

Comment:
It doesn't contain the commit itself, but the hash value of it.

```suggestion
/// Optionally, this may also contain the hash value of the Git commit from
/// which Typst was built. However, this field may be unpopulated.
```